### PR TITLE
[Merged by Bors] - feat: rewrite the copyright header check in Lean

### PR DESCRIPTION
**Pull request by grunweg**
This is one in a series of PRs rewriting most checks from `lint-style.py` in Lean.
This PR moves the check for copyright headers and authors: the check has been rewritten anew (and is hopefully simpler to understand now); this has made it stricter in various places. The necessary adjustments have been made in #13212 and #13247.

This PR also adds a `nolints` file for text-based linters.

--------

This linter is slightly slower than the Python linter: in local testing,
- I believe the Python linter took about 0.6s
- running just `lake exe lint_style` with a no-op linter (incrementing a counter on each file) takes 0.4s
(`time lake exe lint_style` reports `real	0m0,404s; user	0m0,254s; sys	0m0,070s`; this is the second time when no re-compilation is needed)
- with the copyright linter, it takes 0.8s
Any insight into minimising the no-op overhead is welcome (but almost surely out of scope for this PR).

### CODE REVIEW

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  let all_output := (Array.map (fun lint ↦
+    (Array.map (fun (e, n) ↦ ErrorContext.mk e n path)) (lint lines))) all_linters
+  -- XXX: this list is currently not sorted: for github, that's probably fine
+  let errors := Array.flatten all_output
+  formatErrors errors
+  return errors.size > 0
+
+/-- Lint all files referenced in a given import-only file.
+Return the number of files which had new style errors. -/
+def lint_all_files (path : System.FilePath) : IO UInt32 := do
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/TextBased.lean:161]: nit: this, and most of the other names in the file, do not match the usual naming conventions.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  -- Read all module names in Mathlib from the file at `path`.
+  let allModules ← IO.FS.lines path
+  let mut number_error_files := 0
+  for module in allModules do
+    let module := module.stripPrefix "import "
+    -- Exclude `Archive.Sensitivity` and `Mathlib.Tactic.Linter` for now.
+    -- FUTURE: replace this by proper parsing of style exceptions.
+    if #["Archive.Sensitivity", "Mathlib.Tactic.Linter"].contains module then
+      continue
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/TextBased.lean:170]: I'd argue that having support for the style exceptions in nolints is a prerequisite for rewriting these linters.

Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean:170]: In principle, I wholeheartedly agree: indeed, I have implemented parsing style exceptions and updating their exceptions file. I am happy to add it to this PR, if you prefer. I origininally did not, to makes this PR land faster.

I think parsing style exceptions or updating them is essential for rewriting *this* linter:
- there's no reason new copyright exceptions should get added
(This is different from e.g. the large file linter `ERR_NUM_LIN`, where exceptions are an accepted transient state.)
- all existing copyright headers pass the Python lint
- these hard-coded exceptions are one-off cases where the rewritten linter is too strict
Both of them are not hard to fix (but make the code more complicated): do you really care to get this fixed now?

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean]: Indeed, you're right. Addressed.

File: scripts/lint-style.sh [hunk 3]
```diff
@@ -41,6 +41,8 @@ git ls-files 'Mathlib/*.lean' | xargs ./scripts/lint-style.py "$@"
 git ls-files 'Archive/*.lean' | xargs ./scripts/lint-style.py "$@"
 git ls-files 'Counterexamples/*.lean' | xargs ./scripts/lint-style.py "$@"
 
+# Also call the in-progress Lean rewrite of these Python lints: do this in the build step.
```
[the remark this answers is not in the document]
Comment (grunweg) [reply scripts/lint-style.sh:44]: Upon second thought, the step only need the `lint_style` executable; adding that is not too bad.

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean]: Fair enough, I have implemented this now.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors d+

**Comment by grunweg** [comment]
Thanks for the review! Since CI passes now, let me bors it already:
bors r+

**Pull request closed without merging**